### PR TITLE
7738.master

### PR DESCRIPTION
**Pull request by ivandudardb**
Extended the default registration form with custom fields.
Created a .install file with database configuration.
Configured data storage from custom fields into custom database tables.
In the Custom Weather module, transferred the array of cities from the form to a service to utilize it in the registration module.

### CODE REVIEW

**Review by HeorhiiBoliuk — APPROVED** [review]
LGTM

File: web/modules/custom/custom_registration/src/Form/CustomRegistrationForm.php [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+      $container->get('country_manager'),
+      $container->get('entity_type.bundle.info'),
+      $container->get('datetime.time'),
+    );
+  }
+
+  /**
+   * Return array with countries.
+   */
+  public function getCountry(): array {
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_registration/src/Form/CustomRegistrationForm.php:56]: Whay country?

File: web/modules/custom/custom_registration/src/Form/CustomRegistrationForm.php [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    return $countries[] = $this->countryManager->getStandardList();
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_registration/src/Form/CustomRegistrationForm.php:57]: Not sure this is right
Do you understand what you are returning?

File: web/modules/custom/custom_registration/src/Form/CustomRegistrationForm.php [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function form(array $form, FormStateInterface $form_state): array {
+    $countries = $this->getCountry();
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_registration/src/Form/CustomRegistrationForm.php:64]: Country? Countries? decide

File: web/modules/custom/custom_registration/src/Form/CustomRegistrationForm.php [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    $vocabularyId = 'news';
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_registration/src/Form/CustomRegistrationForm.php:65]: unnecessary variable

File: web/modules/custom/custom_registration/src/Form/CustomRegistrationForm.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    $terms = $this->entityTypeManager->getStorage('taxonomy_term')->loadTree($vocabularyId);
+    $options = [];
+    foreach ($terms as $term) {
+      $options[$term->tid] = $term->name;
+    }
+    $form = parent::form($form, $form_state);
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_registration/src/Form/CustomRegistrationForm.php:71]: weird place for parent call

File: web/modules/custom/custom_registration/src/Form/CustomRegistrationForm.php [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+      '#type' => 'select',
+      '#title' => $this->t('City'),
+      '#required' => TRUE,
+      '#options' => $this->userCityHandler->cities(),
+    ];
+
+    $form['interested_in'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Interested In'),
+      '#options' => $options,
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_registration/src/Form/CustomRegistrationForm.php:89]: You have 3 select field, and options for each passed slightly differently for no reason it seems.

File: web/modules/custom/custom_registration/src/Plugin/Block/CustomRegistration.php [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+      $container->get('current_user'),
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function build(): array {
+    $uid = $this->accountProxy->id();
+    if ($uid == 0) {
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_registration/src/Plugin/Block/CustomRegistration.php:53]: ->isAnonymous
Basically same but slightly more clear

File: web/modules/custom/custom_registration/src/Service/DataBaseService.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+use Drupal\Core\Database\Connection;
+
+/**
+ * Saves data from custom registration fields into the database.
+ */
+class DataBaseService {
+
+  /**
+   * Constructs a new DataBaseService object.
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_registration/src/Service/DataBaseService.php:13]: So this is database service, that accepts database service as dependency?

File: web/modules/custom/custom_registration/src/Service/DataBaseService.php [hunk 9]
```diff
@@ -0,0 +1,57 @@
+<?php
+
+namespace Drupal\custom_registration\Service;
+
+use Drupal\Core\Database\Connection;
+
+/**
+ * Saves data from custom registration fields into the database.
+ */
+class DataBaseService {
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_registration/src/Service/DataBaseService.php:10]: Drupal already have service for working with database

File: web/modules/custom/custom_registration/src/Service/DataBaseService.php [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  public function saveData(string $country_value, string $city_value, int $interested_in_value, int $user_id): void {
+    $this->connection->insert('custom_registration_data')
+      ->fields([
+        'uid' => $user_id,
+        'country' => $country_value,
+        'interested_in' => $interested_in_value,
+      ])
+      ->execute();
+
+    $this->connection->merge('custom_weather_data')
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_registration/src/Service/DataBaseService.php:48]: Why city is in a different table?

File: web/modules/custom/custom_weather/src/Service/UserCityHandler.php [hunk 11]
```diff
@@ -6,11 +6,13 @@
 use Drupal\Core\Database\Connection;
 use Drupal\Core\Http\ClientFactory;
 use Drupal\Core\Session\AccountProxyInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
 
 /**
  * Saves the city for displaying weather.
  */
 class UserCityHandler {
+  use StringTranslationTrait;
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_weather/src/Service/UserCityHandler.php:15]: You have trait, but don't have accepted string translation service

**Review by lavrenchuk-dev — CHANGES_REQUESTED** [review]

File: web/modules/custom/custom_registration/src/Form/CustomRegistrationForm.php (hunk 2 above)
Comment (ivandudardb) [reply web/modules/custom/custom_registration/src/Form/CustomRegistrationForm.php:57]: An array of countries. I use it to create a dropdown list.

File: web/modules/custom/custom_weather/src/Service/UserCityHandler.php [hunk 12]
```diff
@@ -85,19 +83,19 @@ public function saveUserCity(string $selected_user_city): void {
...
+      'Lutsk' => t('Lutsk'),
+      'Zhytomyr' => t('Zhytomyr'),
+      'Chernivtsi' => t('Chernivtsi'),
+      'Ternopil' => t('Ternopil'),
+      'Khmelnytskyi' => t('Khmelnytskyi'),
+      'Uzhhorod' => t('Uzhhorod'),
+      'Vinnytsia' => t('Vinnytsia'),
+      'Cherkasy' => t('Cherkasy'),
+      'Poltava' => t('Poltava'),
+      'Chernihiv' => t('Chernihiv'),
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_weather/src/Service/UserCityHandler.php:98]: DI?

**Review by lavrenchuk-dev — CHANGES_REQUESTED** [review]

File: web/modules/custom/custom_weather/src/Service/UserCityHandler.php [hunk 13]
```diff
@@ -19,12 +22,18 @@ class UserCityHandler {
...
+   * @param \Drupal\Core\StringTranslation\TranslationInterface $translation
+   *   The translation service.
    */
   public function __construct(
     protected Connection $connection,
     protected AccountProxyInterface $currentUser,
     protected ClientFactory $httpClient,
+    protected TranslationInterface $translation,
   ) {
+    $this->stringTranslation = $translation;
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_weather/src/Service/UserCityHandler.php:36]: Constructor promotion

**Review by lavrenchuk-dev — CHANGES_REQUESTED** [review]

**Pull request merged**